src/client/line.h

```diff
@@ -0,0 +1,144 @@
+/*
+Minetest
+Copyright (C) 2023 Andrey2470T, AndreyT <andreyt2203@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#pragma once
+
+#include "line_params.h"
+#include <S3DVertex.h>
+#include <SMaterial.h>
+#include <ISceneNode.h>
+#include "clientenvironment.h"
+#include <ISceneManager.h>
+#include <unordered_map>
+#include "clientevent.h"
+#include "log.h"
+
+class LineSceneNode;
+
+class LineSceneNodeManager
+{
+public:
+	LineSceneNodeManager(ClientEnvironment* env)
+	: m_env(env)
+	{
+
+	}
+
+	~LineSceneNodeManager()
+	{
+		for (auto &p : m_line_scenenode_list)
+			remove3DLineFromList(p.first);
```

Comment:
This creates problem when exiting to main menu. The iterator fails to point after an element is removed.

My suggestion:
```diff
diff --git a/src/client/line.cpp b/src/client/line.cpp
index a60f9efe4..84345d607 100644
--- a/src/client/line.cpp
+++ b/src/client/line.cpp
@@ -405,12 +405,17 @@ void LineSceneNodeManager::add3DLineToList(u32 id, LineSceneNode *line_node)
        m_line_scenenode_list.insert({id, line_node});
 }

+void LineSceneNodeManager::remove3DLineById(u32 id)
+{
+       m_line_scenenode_list[id]->remove();
+}
+
 void LineSceneNodeManager::remove3DLineFromList(u32 id)
 {
        if (!m_line_scenenode_list[id])
                return;

-       m_line_scenenode_list[id]->remove();
+       remove3DLineById(id);
        m_line_scenenode_list.erase(id);
 }

diff --git a/src/client/line.h b/src/client/line.h
index bd084b38a..e1d347c90 100644
--- a/src/client/line.h
+++ b/src/client/line.h
@@ -43,10 +43,11 @@ class LineSceneNodeManager
        ~LineSceneNodeManager()
        {
                for (auto &p : m_line_scenenode_list)
-                       remove3DLineFromList(p.first);
+                       remove3DLineById(p.first);
        }

        void add3DLineToList(u32 id, LineSceneNode *line_node);
+       void remove3DLineById(u32 id);
        void remove3DLineFromList(u32 id);
        void handle3DLineEvent(ClientEventType event_type, u32 id, LineParams *params);
        void step(f32 dtime);
```

Reply:
I added just the only `p.second->remove();` line in the cycle without that method.